File: 02-FZHDeleteCells/02-FZHDeleteCells(OC)/02-FZHDeleteCells(OC)/ViewController.m
//
//  ViewController.m
//  02-FZHDeleteCells(OC)
//
//  Created by 冯志浩 on 16/8/9.
//  Copyright © 2016年 FZH. All rights reserved.
//

#import "ViewController.h"
#define SCREEN_WIDTH  [UIScreen mainScreen].bounds.size.width
#define SCREEN_HEIGHT [UIScreen mainScreen].bounds.size.height
@interface ViewController ()<UITableViewDelegate,UITableViewDataSource>
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) NSMutableArray *dataSource;
@property (nonatomic, strong) NSMutableArray *deleteArray;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupNavigationItem];
    [self setupTableView];
    [self initData];
    
}

- (void)initData{
    self.dataSource = [NSMutableArray arrayWithObjects:@"1",@"2",@"3",@"4",@"5",@"6",@"7",@"8", nil];
    self.deleteArray = [NSMutableArray arrayWithObjects:@"", nil];
}

- (void)setupTableView{
    self.tableView = [[UITableView alloc]initWithFrame:CGRectMake(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT) style:UITableViewStylePlain];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    [self.view addSubview:self.tableView];
}

- (void)setupNavigationItem{

    UIBarButtonItem *editBtn = [[UIBarButtonItem alloc]initWithTitle:@"编辑" style:UIBarButtonItemStyleDone target:self action:@selector(buttonDidTouch:)];
    UIBarButtonItem *doneBtn = [[UIBarButtonItem alloc]initWithTitle:@"完成" style:UIBarButtonItemStyleDone target:self action:@selector(buttonDidTouch:)];
    UIBarButtonItem *deleteBtn = [[UIBarButtonItem alloc]initWithTitle:@"删除" style:UIBarButtonItemStyleDone target:self action:@selector(buttonDidTouch:)];
    UIBarButtonItem *selectAllBtn = [[UIBarButtonItem alloc]initWithTitle:@"全选" style:UIBarButtonItemStyleDone target:self action:@selector(buttonDidTouch:)];
    
    editBtn.tag = 1001;
    doneBtn.tag = 1002;
    deleteBtn.tag = 1003;
    self.navigationItem.leftBarButtonItems = @[editBtn,doneBtn,deleteBtn,selectAllBtn];
}

- (void)buttonDidTouch:(UIButton *)btn{
    if (btn.tag == 1001) {//edit
        self.tableView.editing = YES;
        self.tableView.allowsMultipleSelectionDuringEditing = YES;
    }else if (btn.tag == 1002){//done
        self.tableView.editing = NO;
    }else if (btn.tag == 1003){//delete
        [self.dataSource removeObjectsInArray:self.deleteArray];
        [self.tableView reloadData];
    }else{
        if (self.tableView.editing == NO) {
            return;
        }else{
            for (int i = 0; i < self.dataSource.count; i++) {
                NSIndexPath *indexPath = [NSIndexPath indexPathForRow:i inSection:0];
                [self.tableView selectRowAtIndexPath:indexPath animated:YES scrollPosition:UITableViewScrollPositionNone];
            }
            [self.deleteArray addObjectsFromArray:self.dataSource];
        }
    }
}
#pragma mark - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.dataSource.count;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    static NSString * cellID = @"cellID";
    UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:cellID];
    if (!cell) {
        cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellID];
    }
    cell.textLabel.text = self.dataSource[indexPath.row];
    return cell;
}
#pragma mark - UITableViewDelegate
//选中
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [self.deleteArray addObject:self.dataSource[indexPath.row]];
}
//取消选中
- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath{
    [self.deleteArray removeObject:self.dataSource[indexPath.row]];
}
- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath{

    return UITableViewCellEditingStyleDelete|UITableViewCellEditingStyleInsert;
}
@end

File: 02-FZHDeleteCells/02-FZHDeleteCells(Swift)/02-FZHDeleteCells(Swift)/ViewController.swift
//
//  ViewController.swift
//  02-FZHDeleteCells(Swift)
//
//  Created by 冯志浩 on 16/8/8.
//  Copyright © 2016年 FZH. All rights reserved.
//

import UIKit

class ViewController: UIViewController,UITableViewDelegate,UITableViewDataSource {
    let SCREEN_WIDTH = UIScreen.main.bounds.size.width
    let SCREEN_HEIGHT = UIScreen.main.bounds.size.height
    let tableView = UITableView()
    var dataSource: NSMutableArray = ["1","2","3","4","5","6","7","8"]
    var deleteArr: NSMutableArray = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupNavigationBar()
        setupTableView()
    }
    
    func setupNavigationBar() -> Void {
        let editButton = UIBarButtonItem.init(title: "编辑", style: .done, target: self, action: #selector(buttonDidTouch))
        let doneButton = UIBarButtonItem.init(title: "完成", style: .done, target: self, action: #selector(buttonDidTouch))
        let deleteButton = UIBarButtonItem.init(title: "删除", style: .done, target: self, action: #selector(buttonDidTouch))
        let selectAllButton = UIBarButtonItem.init(title: "全选", style: .done, target: self, action: #selector(buttonDidTouch))
        
        editButton.tag = 1001
        doneButton.tag = 1002
        deleteButton.tag = 1003
        
        self.navigationItem.leftBarButtonItems = [editButton,doneButton,deleteButton,selectAllButton]
    }
    
    func buttonDidTouch(_ btn:UIButton) -> Void {
        if btn.tag == 1001 {//编辑
            tableView.isEditing = true
            tableView.allowsMultipleSelectionDuringEditing = true
        }else if btn.tag == 1002{//完成
            tableView.isEditing = false
        }else if btn.tag == 1003{//删除
            self.dataSource.removeObjects(in: self.deleteArr as [AnyObject])
            tableView.reloadData()
        }else{//全选
            if tableView.isEditing == false {//如果处于完成状态则直接返回
                return
            }
            for row in 0...self.dataSource.count {
                let indexPath = IndexPath.init(row: row, section: 0)
                tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
            }
            self.deleteArr.addObjects(from: self.dataSource as [AnyObject])
        }
    }
    
    func setupTableView() -> Void {
        tableView.frame = CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_HEIGHT)
        tableView.delegate = self
        tableView.dataSource = self
        self.view.addSubview(tableView)
    }
//    MARK: UITableViewDataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.dataSource.count
    }
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell.init(style: .default, reuseIdentifier: "UITableViewCell")
        cell.textLabel?.text = self.dataSource[(indexPath as NSIndexPath).row] as? String
        return cell
        
    }
    
//MARK: UITableViewDelegate
//    设置编辑style
    func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return UITableViewCellEditingStyle.init(rawValue: UITableViewCellEditingStyle.delete.rawValue | UITableViewCellEditingStyle.insert.rawValue)!
    }
//    选中则添加到删除数组中
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.deleteArr.add(self.dataSource[(indexPath as NSIndexPath).row])
    }
//    取消选中在数组中删除
    func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        self.deleteArr.remove(self.dataSource[(indexPath as NSIndexPath).row])
    }
}

